### PR TITLE
[GL] Polygon: Impl. of copy constructor and assignment operator

### PR DESCRIPTION
**Pull request by TomFischer**
As titled.

Fixes issue #1155.

**Update (2016-04-25)**
Removed the assignment operator since it isn't used anywhere and marked it explicitly as deleted.

### CODE REVIEW

File: GeoLib/Polygon.cpp [hunk 1]
```diff
@@ -417,27 +417,25 @@ void Polygon::splitPolygonAtPoint (std::list<GeoLib::Polygon*>::iterator polygon
...
-			std::list<GeoLib::Polygon*>::iterator polygon0_it, polygon1_it;
-			polygon1_it =
-			        _simple_polygon_list.insert (_simple_polygon_list.erase (
-			                                             polygon_it), polygon1);
-			polygon0_it = _simple_polygon_list.insert (polygon1_it, polygon0);
-
-			splitPolygonAtPoint (polygon0_it);
-			splitPolygonAtPoint (polygon1_it);
+			auto polygon1_it = _simple_polygon_list.insert(
+			    _simple_polygon_list.erase(polygon_it), new Polygon(polyline1));
```
Comment (chleh) [inline GeoLib/Polygon.cpp]: You erase a pointer to a polygon from `_simple_polygon_list`. Does that leak memory?
Just a note: It seems like `*polygon_it = new Polygon(polyline1);` is equivalent to the given statement.

File: GeoLib/Polygon.cpp [hunk 2]
```diff
@@ -38,6 +38,25 @@ Polygon::Polygon(Polygon const& other)
 	}
 }
 
+Polygon& Polygon::operator=(Polygon const& other)
+{
+	if (other._ply_pnts != _ply_pnts) {
+		ERR(
+		    "Polygon::operator=: polygon other and this polygon are not based "
+		    "on the same point vector. Aborting");
```
Comment (chleh) [inline GeoLib/Polygon.cpp]: Why is that necessary?

File: GeoLib/Polygon.cpp [hunk 3]
```diff
@@ -38,6 +38,25 @@ Polygon::Polygon(Polygon const& other)
...
+		std::abort();
+	}
+	if (other == *this)
```
Comment (chleh) [inline GeoLib/Polygon.cpp]: You compare for equality (i.e., compare the objects), not for identity (i.e., compare their addresses), here. I'd compare `&other == this`. That should also be faster.

File: GeoLib/Polygon.cpp [hunk 4]
```diff
@@ -38,6 +38,25 @@ Polygon::Polygon(Polygon const& other)
...
+		return *this;
+
+	Polygon tmp(other);
+	std::swap(tmp._aabb, _aabb);
+	std::swap(tmp._simple_polygon_list, _simple_polygon_list);
+	std::swap(tmp._length, _length);
+	std::swap(tmp._ply_pnt_ids, _ply_pnt_ids);
```
Comment (chleh) [inline GeoLib/Polygon.cpp]: You only have two data members inside `Polygon`: `_simple_polygon_list` and `_aabb`. Are the other two defined in the base class?
Then it would be better to use `operator=` of `Polyline` for them, e.g., `Polyline::operator=(*this, other);` (I'm not sure if it works like this.) If you think, the code becomes too complicated by that, just leave it as it is.

File: Tests/GeoLib/TestPolygon.cpp [hunk 5]
```diff
@@ -199,6 +199,28 @@ TEST_F(PolygonTest, isPolylineInPolygon)
...
+	// create a polygon
+	GeoLib::Polyline polyline(_pnts);
+	polyline.addPoint(0);
+	polyline.addPoint(1);
+	polyline.addPoint(2);
+	polyline.addPoint(0);
+	GeoLib::Polygon polygon(polyline);
+	// assign an other polygon
+	polygon = *_polygon;
+	// ToDo change to segments
```
Comment (chleh) [inline Tests/GeoLib/TestPolygon.cpp]: Very minor thing: I suggest to agree on one spelling for TODO. E.g., my IDE recognizes `TODO` and gathers all those lines in a nice list.

**Comment by chleh** [comment]
Many comments, nothing big. Afterwards :+1:

File: GeoLib/Polygon.cpp (hunk 2 above)
Comment (endJunction) [inline GeoLib/Polygon.cpp]: The polygons do not store the points but have a reference to point vector. So depending on how the points are managed all can go wrong here w/o the if. The details need to be explained anyway; Tom?

File: GeoLib/Polygon.cpp [hunk 6]
```diff
@@ -417,27 +444,25 @@ void Polygon::splitPolygonAtPoint (std::list<GeoLib::Polygon*>::iterator polygon
...
 			// create two closed polylines
-			GeoLib::Polygon* polygon0 (new GeoLib::Polygon(*(*polygon_it)));
+			GeoLib::Polyline polyline0{*(*polygon_it)};
 			for (std::size_t k(0); k <= idx0; k++)
-				polygon0->addPoint ((*polygon_it)->getPointID (k));
-			for (std::size_t k(idx1 + 1); k < (*polygon_it)->getNumberOfPoints(); k++)
-				polygon0->addPoint ((*polygon_it)->getPointID (k));
-			polygon0->initialise();
+				polyline0.addPoint((*polygon_it)->getPointID(k));
+			for (std::size_t k(idx1 + 1);
```
Comment (endJunction) [inline GeoLib/Polygon.cpp]: Declaration shadows a local variable `k`; see line 402. Maybe not a bug but could be nicer.

File: GeoLib/Polygon.cpp [hunk 7]
```diff
@@ -417,27 +444,25 @@ void Polygon::splitPolygonAtPoint (std::list<GeoLib::Polygon*>::iterator polygon
...
+			     k < (*polygon_it)->getNumberOfPoints(); k++)
+				polyline0.addPoint((*polygon_it)->getPointID(k));
 
-			GeoLib::Polygon* polygon1 (new GeoLib::Polygon(*(*polygon_it)));
+			GeoLib::Polyline polyline1{*(*polygon_it)};
 			for (std::size_t k(idx0); k <= idx1; k++)
```
Comment (endJunction) [inline GeoLib/Polygon.cpp]: Same here...

File: GeoLib/Polygon.cpp [hunk 8]
```diff
@@ -30,6 +30,33 @@ Polygon::Polygon(const Polyline &ply, bool init) :
...
+		ERR(
+		    "Polygon::operator=: polygon other and this polygon are not based "
+		    "on the same point vector. Aborting");
+		std::abort();
+	}
+	if (other == *this)
+		return *this;
+
+	Polygon tmp(other);
+	std::swap(tmp._aabb, _aabb);
```
Comment (endJunction) [inline GeoLib/Polygon.cpp]: This can be handled in another PR but the precedence is created here:
`warning: definition of implicit copy assignment operator for 'AABB' is deprecated because it has a user-declared copy constructor [-Wdeprecated]`
... requested in std::swap<GeoLib::AABB> at this line.

File: GeoLib/Polygon.cpp (hunk 2 above)
Comment (chleh) [inline GeoLib/Polygon.cpp]: I just saw that `const std::vector<Point*> &_ply_pnts;` is a reference member of `Polyline`. Therefore it cannot be reset by the assignment operator.
Different thing: However, it would probably be cheaper to compare the addresses and not the whole vectors. Furthermore the check `other == *this` should probably be done first.

File: GeoLib/Polygon.cpp (hunk 1 above)
Comment (endJunction) [inline GeoLib/Polygon.cpp]: If this code is tested, there is no memleak reported by address sanitizer...

Comment (TomFischer) [inline GeoLib/Polygon.cpp]: As far as I see the code isn't tested. At the moment I think @chleh is right. This is probably a memory leak.

File: GeoLib/Polygon.cpp [hunk 9]
```diff
@@ -72,7 +72,7 @@ bool Polygon::isPntInPolygon (GeoLib::Point const & pnt) const
 	std::size_t n_intersections (0);
 	GeoLib::Point s;
 
-	if (_simple_polygon_list.empty ()) {
+	if (_simple_polygon_list.front() == this) {
```
Comment (chleh) [inline GeoLib/Polygon.cpp]: There are two occasions of `_simple_polygon_list.empty()` left in Polygon.cpp. I guess, they also have to be changed.

**Comment by chleh** [comment]
One comment, afterwards :+1:

**Comment by endJunction** [comment]
:fast_forward:
Please make sure after a rebase that all tests pass (it takes sometimes some time).

**Comment by TomFischer** [comment]
At my laptop all tests pass after rebase. Unfortunately I can not see why Jenkins fails because I am not at the UFZ and thus have no access to Jenkins. If somebody see that the failing Jenkins is not caused by my changes feel free to merge.

**Comment by norihiro-w** [comment]
@TomFischer you can go to the Jenkins via https://my.ufz.de/login

**Comment by TomFischer** [comment]
> @TomFischer you can go to the Jenkins via https://my.ufz.de/login

Thanks for the info. Will try it.

**Comment by bilke** [comment]
@TomFischer I played around with some new Jenkins functionality. That is why there is another Jenkins build run (called `Jenkins` in the check list below) is there. I have merged #1170 which should fix this new Jenkins run. The `default` Jenkins job builds fine.

**Comment by TomFischer** [comment]
@bilke Thanks for the info.
